pennylane/measurements.py

```diff
@@ -587,3 +593,115 @@ def circuit():
     """
     # pylint: disable=protected-access
     return MeasurementProcess(State, wires=qml.wires.Wires(wires))
+
+
+T = TypeVar("T")
+
+
+class MeasurementValueError(ValueError):
+    """Error raised when an unknown measurement value is being used."""
+
+
+class MeasurementValue(Generic[T]):
+    """A class representing unknown measurement outcomes in the qubit model.
+
+    By default, measurements in the computational basis are assumed.
+
+    Args:
+        measurement_id (str): The id of the measurement that this object depends on.
+        zero_case (float): the measurement outcome in case the first
+            measurement outcome was obtained
+        one_case (float): the measurement outcome in case the second
+            measurement outcome was obtained
+    """
+
+    __slots__ = ("_depends_on", "_zero_case", "_one_case", "_control_value")
+
+    def __init__(
+        self,
+        measurement_id: str,
+        zero_case: float = 0,
+        one_case: float = 1,
+    ):
+        self._depends_on = measurement_id
+        self._zero_case = zero_case
+        self._one_case = one_case
+        self._control_value = one_case  # By default, control on the one case
+
+    @property
+    def branches(self):
+        """A dictionary representing all the possible outcomes of the MeasurementValue."""
+        branch_dict = {}
+        branch_dict[(0,)] = self._zero_case
+        branch_dict[(1,)] = self._one_case
+        return branch_dict
+
+    def __eq__(self, control_value):
+        """Allow asserting measurement values."""
+        measurement_outcomes = {self._zero_case, self._one_case}
+        if control_value not in measurement_outcomes:
+            raise MeasurementValueError(
+                f"Unknown measurement value asserted; the set of possible measurement outcomes is: {measurement_outcomes}."
+            )
+
+        self._control_value = control_value
+        return self
+
+    @property
+    def control_value(self):
+        """The control value to consider for the measurement outcome."""
+        return self._control_value
+
+    @property
+    def measurements(self):
+        """List of all measurements this MeasurementValue depends on."""
+        return [self._depends_on]
```

Comment:
Will this always be a length-1 list?

Reply:
At the moment, yes, but we'd like to introduce more complex conditional structures such as `qml.cond(m0 + m1 == 2, RY)(0.1234, wires=2)`. In such a case, this list would have 2 items.